packages/dd-trace/src/scope/base.js

```diff
@@ -136,59 +140,87 @@ class Scope {
   }
 }
 
+function wrapThen (then, scope, span) {
+  return function thenWithTrace (onFulfilled, onRejected) {
+    const args = new Array(arguments.length)
+
+    for (let i = 0, l = args.length; i < l; i++) {
+      args[i] = scope.bind(arguments[i], span)
+    }
+
+    return then.apply(this, args)
+  }
+}
+
 function wrapAddListener (addListener, scope, span) {
   return function addListenerWithTrace (eventName, listener) {
-    if (!listener || listener._datadog_bound) return addListener.apply(this, arguments)
+    if (!listener || listener._datadog_unbound || listener.listener) {
+      return addListener.apply(this, arguments)
+    }
 
     const bound = scope.bind(listener, scope._spanOrActive(span))
 
-    bound._datadog_bound = true
-    listener._datadog_id = listener._datadog_id || ++id
+    this._datadog_events = this._datadog_events || {}
 
     if (!this._datadog_events[eventName]) {
-      this._datadog_events[eventName] = {}
+      this._datadog_events[eventName] = new WeakMap()
     }
 
     const events = this._datadog_events[eventName]
 
-    if (!events[listener._datadog_id]) {
-      events[listener._datadog_id] = []
+    if (!events.has(listener)) {
+      events.set(listener, [])
     }
 
-    events[listener._datadog_id].push(bound)
+    events.get(listener).push(bound)
 
     return addListener.call(this, eventName, bound)
   }
 }
 
 function wrapRemoveListener (removeListener) {
   return function removeListenerWithTrace (eventName, listener) {
-    const listeners = this._datadog_events[eventName]
+    const listeners = this._datadog_events && this._datadog_events[eventName]
 
-    if (!listener || !listeners || !listeners[listener._datadog_id]) {
+    if (!listener || !listeners || !listeners.has(listener)) {
       return removeListener.apply(this, arguments)
     }
 
-    let bound
-
-    while ((bound = listeners[listener._datadog_id].pop())) {
+    for (const bound of listeners.get(listener)) {
       removeListener.call(this, eventName, bound)
     }
 
+    listeners.delete(listener)
+
     return removeListener.call(this, eventName, listener)
   }
 }
 
 function wrapRemoveAllListeners (removeAllListeners) {
   return function removeAllListenersWithTrace (eventName) {
-    if (eventName) {
-      this._datadog_events[eventName] = {}
+    if (this._datadog_events && eventName) {
+      delete this._datadog_events[eventName]
```

Comment:
Let's put a check for `this._datadog_events` outside this `if` since `delete`s are expensive.